File: brick.php
---
layout: null
---
<?php

/*
 * Brick. Webfonts that actually look good
 */

// Catalogue array
{% assign fonts = site.fonts %}
$catalogue = array(
    {% for font in fonts %}
    {% if font.layout == "font" %}
    "{{ font.family }}" => array(
        {% for style in font.styles %}
        "{{ style[0] }}" => "{{ style[1] }}"{% unless forloop.last %},{% endunless %}
        {% endfor %}
    ){% unless forloop.last %},{% endunless %}
    {% endif %}
    {% endfor %}
);

// Headers
header('Content-type: text/css');
header('X-Content-Type-Options: nosniff');
header('Cache-Control: public, max-age=2628000');
header('Expires: ' . gmdate('D, d M Y H:i:s \G\M\T', time() + 2628000));
header('Last-Modified: ' . gmdate('D, d M Y H:i:s \G\M\T', filemtime(__FILE__)));
header('Pragma: Public');

// Template
$BASE = "@font-face{font-family:'%s';font-style:%s;font-weight:%s;src:%s}";
$SVG = "@media screen and (-webkit-min-device-pixel-ratio:0){@font-face{font-family:'%s';font-style:%s;font-weight:%s;src:%s}}";

$query = explode("/", preg_replace("/\/$|^\//", "", urldecode($_SERVER['REQUEST_URI'])));

if ($query[0] == "") {
    header($_SERVER['SERVER_PROTOCOL'] . ' 400 Bad Request');
    exit();
}

foreach ($query as $key=>$val) {
    // Query sections
    $val = explode(":", $val);
    $family = str_replace("+", "", $val[0]);
    $weights = explode(",", $val[1]);
    $flags = isset($val[2]) ? $val[2] : '';

    foreach ($weights as $weight) {
        // Build font URL
        $woff = "//brick.a.ssl.fastly.net/fonts/"
              . strtolower(str_replace(" ", '', $family))
              . "/"
              . $weight
              . ".woff";

        $svg = "//brick.a.ssl.fastly.net/fonts/"
              . strtolower(str_replace(" ", '', $family))
              . "/"
              . $weight
              . ".svg"
              . "#"
              . strtolower(str_replace(" ", '', $family));


        // Start with no URI's
        $uri = '';

        // Process flags
        if (strpos($flags, 'f') === false) {
            $local = $catalogue[$family][$weight];
            $uri .= "local('" . $local . "'),";
        }

        // Add font URL
        $uri .= "url(" . $woff . ") format('woff')";
        
        if (substr($weight, -1) == "i") {
            $style = 'italic';
            $weight = substr($weight, 0, -1);
        } else {
            $style = 'normal';
        }

        echo sprintf($BASE, $family, $style, $weight, $uri);
        if (strpos($flags, 's') !== false) {
            echo sprintf($SVG, $family, $style, $weight, "url(" . $svg . ") format('svg')");
        }
        
    }
}

?>

File: brick.go
---
---

// Brick. Webfonts that actually look good

package main

import (
	"bytes"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// Handles an incoming CSS request, and builds CSS based on query parameters
// (font family, weights and styles, flags)
func handler(w http.ResponseWriter, r *http.Request) {
	{% assign fonts = site.fonts %}
	var fonts = map[string](map[string]string) {
		{% for font in fonts %}
		"{{ font.family }}": map[string]string {
			{% for style in font.styles %}
			"{{ style[0] }}": "{{ style[1] }}",
			{% endfor %}
		},
		{% endfor %}
	}

	w.Header().Set("Content-type", "text/css")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.Header().Set("Cache-Control", "public, max-age=2628000")
	w.Header().Set("Expires", time.Now().Add(2628000*time.Second).Format(time.RFC1123))
	w.Header().Set("Last-Modified", "{{ site.time | date: '%a, %d %b %Y %T %Z'}}")
	w.Header().Set("Pragma", "Public")
	w.Header().Set("Server", "Brick")

	query := strings.Split(strings.Trim(r.URL.Path, "/"), "/")

	if len(query) == 0 || query[0] == "" {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	for _, font := range query {
		data := strings.Split(font, ":")
		if len(data) < 2 {
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		family := strings.Replace(data[0], "+", " ", -1)
		weights := strings.Split(data[1], ",")

		// Flags
		flags := ""
		if len(data) > 2 {
			flags = data[2]
		}

		for _, weight := range weights {
			// Verify that variant exists, else move on
			if _, exists := fonts[family][weight]; !exists {
				continue
			}

			var uri bytes.Buffer

			// Local font URI
			if !strings.Contains(flags, "f") {
				local := fonts[family][weight]
				uri.WriteString("local('")
				uri.WriteString(local)
				uri.WriteString("'),")
			}

			// Brick URI
			uri.WriteString("url(//brick.a.ssl.fastly.net/fonts/")
			uri.WriteString(strings.ToLower(strings.Replace(family, " ", "", -1)))
			uri.WriteString("/")
			uri.WriteString(weight)
			uri.WriteString(".woff) format('woff')")

			// Real weight and style
			realWeight := strings.TrimSuffix(weight, "i")
			style := "normal"
			if len(realWeight) < len(weight) {
				style = "italic"
			}

			template := "@font-face{font-family:'%s';font-style:%s;font-weight:%s;src:%s}"
			fmt.Fprintf(w, template, family, style, realWeight, uri.String())
		}
	}
}

// Starts an HTTP server to listen for incoming requests
func main() {
	http.HandleFunc("/", handler)
	http.ListenAndServe(":9811", nil)
}
